src/rdma.c

```diff
@@ -0,0 +1,985 @@
+    }
+
+    vk_free(ctx->cmd_buf);
+    ctx->cmd_buf = NULL;
+}
+
+static int rdmaSetupIoBuf(valkeyContext *c, RdmaContext *ctx, struct rdma_cm_id *cm_id) {
+    int access = IBV_ACCESS_LOCAL_WRITE;
+    size_t length = sizeof(ValkeyRdmaCmd) * VALKEY_RDMA_MAX_WQE * 2;
+    ValkeyRdmaCmd *cmd;
+    int i;
+
+    /* setup CMD buf & MR */
+    ctx->cmd_buf = vk_calloc(length, 1);
+    ctx->cmd_mr = ibv_reg_mr(ctx->pd, ctx->cmd_buf, length, access);
+    if (!ctx->cmd_mr) {
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: reg recv mr failed");
+        goto destroy_iobuf;
+    }
+
+    for (i = 0; i < VALKEY_RDMA_MAX_WQE; i++) {
+        cmd = ctx->cmd_buf + i;
+
+        if (rdmaPostRecv(ctx, cm_id, cmd) == VALKEY_ERR) {
+            valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: post recv failed");
+            goto destroy_iobuf;
+        }
+    }
+
+    for (i = VALKEY_RDMA_MAX_WQE; i < VALKEY_RDMA_MAX_WQE * 2; i++) {
+        cmd = ctx->cmd_buf + i;
+        cmd->keepalive.opcode = VALKEY_RDMA_INVALID_OPCODE;
+    }
+
+    /* setup recv buf & MR */
+    access = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE;
+    length = VALKEY_RDMA_DEFAULT_RX_LEN;
+    ctx->recv_buf = vk_calloc(length, 1);
+    ctx->recv_length = length;
+    ctx->recv_mr = ibv_reg_mr(ctx->pd, ctx->recv_buf, length, access);
+    if (!ctx->recv_mr) {
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: reg send mr failed");
+        goto destroy_iobuf;
+    }
+
+    return VALKEY_OK;
+
+destroy_iobuf:
+    rdmaDestroyIoBuf(ctx);
+    return VALKEY_ERR;
+}
+
+static int rdmaAdjustSendbuf(valkeyContext *c, RdmaContext *ctx, unsigned int length) {
+    int access = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE;
+
+    if (length == ctx->send_length) {
+        return VALKEY_OK;
+    }
+
+    /* try to free old MR & buffer */
+    if (ctx->send_length) {
+        ibv_dereg_mr(ctx->send_mr);
+        vk_free(ctx->send_buf);
+        ctx->send_length = 0;
+    }
+
+    /* create a new buffer & MR */
+    ctx->send_buf = vk_calloc(length, 1);
+    ctx->send_length = length;
+    ctx->send_mr = ibv_reg_mr(ctx->pd, ctx->send_buf, length, access);
+    if (!ctx->send_mr) {
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: reg send buf mr failed");
+        vk_free(ctx->send_buf);
+        ctx->send_buf = NULL;
+        ctx->send_length = 0;
+        return VALKEY_ERR;
+    }
+
+    return VALKEY_OK;
+}
+
+
+static int rdmaSendCommand(RdmaContext *ctx, struct rdma_cm_id *cm_id, ValkeyRdmaCmd *cmd) {
+    struct ibv_send_wr send_wr, *bad_wr;
+    struct ibv_sge sge;
+    ValkeyRdmaCmd *_cmd;
+    int i;
+    int ret;
+
+    /* find an unused cmd buffer */
+    for (i = VALKEY_RDMA_MAX_WQE; i < 2 * VALKEY_RDMA_MAX_WQE; i++) {
+        _cmd = ctx->cmd_buf + i;
+        if (_cmd->keepalive.opcode == VALKEY_RDMA_INVALID_OPCODE) {
+            break;
+        }
+    }
+
+    assert(i < 2 * VALKEY_RDMA_MAX_WQE);
```

Comment:
There is a [Valkey Over RDMA document](https://github.com/valkey-io/valkey-doc/pull/123/files) pending on valkey-doc.

Valkey Over RDMA has control-plane (control messages) and data-plane (payload transfer):
data-plane: use `IBV_WR_RDMA_WRITE_WITH_IMM` frequently which does not consume these sending buffer.
control-plane: use `IBV_WR_SEND` very occasionally. Once a control command completes(usually expected less than 50us), this buffer gets free (please see `static int connRdmaHandleSend(ValkeyRdmaCmd *cmd)`). I believe 1024 commands are inflight state, something gets wrong.

Reply:
Sure, but wouldn't it be better instead to return `VALKEY_ERR`?  Possibly even call `valkeySetError(VALKEY_ERR_OTHER, "No empty RDMA buffers");`

Right now it will just `abort()` if assertions are generated and cause `UB` if this happens with a build that defines `NDEBUG`.